File: app/src/main/java/com/adaca/memoria/GameActivity.java
package com.adaca.memoria;

import static android.view.View.SYSTEM_UI_FLAG_FULLSCREEN;
import static android.view.View.SYSTEM_UI_FLAG_HIDE_NAVIGATION;
import static android.view.View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY;
import static android.view.View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN;
import static android.view.View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION;
import static android.view.View.SYSTEM_UI_FLAG_LAYOUT_STABLE;
import static android.view.WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_SHORT_EDGES;

import android.content.Intent;
import android.graphics.drawable.AnimationDrawable;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.provider.Settings;
import android.view.View;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.res.ResourcesCompat;
import androidx.work.WorkManager;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import model.Game;
import savexport.DatabaseController;
import utils.MediaPlayerManager;
import utils.TimeManager;
import worker.WorkerUtils;

public class GameActivity extends AppCompatActivity {
    private final List<Integer> figures = new ArrayList<Integer>() {
        {
            add((R.drawable.circle));
            add((R.drawable.circle));
            add((R.drawable.hexagon));
            add((R.drawable.hexagon));
            add((R.drawable.losango));
            add((R.drawable.losango));
            add((R.drawable.pentagon));
            add((R.drawable.pentagon));
            add((R.drawable.triangulo));
            add((R.drawable.triangulo));
            add((R.drawable.square));
            add((R.drawable.square));
        }
    };

    private TimeManager timeManager;
    private MediaPlayerManager mediaPlayerManager;
    private TextView points;
    private View mainView;
    private ImageButton firstCard;
    protected int hits, tips, misses;
    protected int missesInARow = 0;
    boolean hit;

    private void init() {
        timeManager = TimeManager.getInstance();
        mediaPlayerManager = MediaPlayerManager.getInstance(getApplication());
        mainView = getWindow().getDecorView();
    }

    protected void onCreate(Bundle paramBundle) {
        super.onCreate(paramBundle);
        setContentView(R.layout.activity_game);

        init();

        timeManager.setStartTime();

        Collections.shuffle(figures);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P)
            getWindow().getAttributes().layoutInDisplayCutoutMode = LAYOUT_IN_DISPLAY_CUTOUT_MODE_SHORT_EDGES;

        points = findViewById(R.id.tvNumPontos);
        points.setText(String.valueOf(hits));
    }

    public void cardClickListener(View imageButton) {
        if (firstCard == null || imageButton.getId() != firstCard.getId()) {
            String label = imageButton.getResources().getResourceEntryName(imageButton.getId());
            int indexBackground = Integer.parseInt(label.split("d")[1]);
            setButtonsClickable(false);

            int figureBackground = figures.get(indexBackground);
            imageButton.setBackgroundResource(figureBackground);
            imageButton.setTag(figureBackground);

            if (firstCard != null) {
                hit = ((Integer) firstCard.getTag() == figureBackground);
            } else {
                firstCard = (ImageButton) imageButton;
            }

            new Handler().postDelayed(() -> {
                imageButton.setBackgroundResource(R.drawable.cartas);
                if (firstCard != imageButton) {
                    if (!hit) {
                        misses++;
                        missesInARow++;
                    } else {
                        mediaPlayerManager.hitFX.start();
                        points.setText(String.valueOf(++hits));
                        missesInARow = 0;
                        firstCard.setVisibility(View.INVISIBLE);
                        imageButton.setVisibility(View.INVISIBLE);
                        hit = false;

                        if (hits == 6) {
                            startActivity(new Intent(this, CongratulationsActivity.class));
                            finish();
                        }
                    }

                    if (missesInARow % 3 == 0 && missesInARow > 0) {
                        showTip(firstCard);
                        tips++;
                    }
                    firstCard = null;
                }
                setButtonsClickable(true);
            }, 500);
        }
    }

    public void onWindowFocusChanged(boolean paramBoolean) {
        super.onWindowFocusChanged(paramBoolean);
        if (paramBoolean)
            mainView.setSystemUiVisibility(SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                    | SYSTEM_UI_FLAG_LAYOUT_STABLE
                    | SYSTEM_UI_FLAG_FULLSCREEN
                    | SYSTEM_UI_FLAG_HIDE_NAVIGATION
                    | SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                    | SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION);
    }

    protected void onStop() {
        super.onStop();
        saveData();
    }

    private void saveData() {
        timeManager.setEndTime();
        Game game = new Game();
        game.setDeviceUID(Settings.Secure.getString(getContentResolver(), "bluetooth_name"));
        game.setInicio(timeManager.getDate());
        game.setDuracao(timeManager.gameTime());
        game.setAcertos(hits);
        game.setDicas(tips);
        game.setErros(misses);
        String resultDB = new DatabaseController(getBaseContext()).insertData(
                game
        );
        WorkManager.getInstance(this).enqueue(WorkerUtils.addDataToRequest((game.toString())));

        Toast.makeText(getApplicationContext(), resultDB, Toast.LENGTH_LONG).show();
    }

    private void showTip(View firstCard) {
        ArrayList<Integer> possibilities = new ArrayList<>();

        for (int index = 0; index < 12; index++) {
            if (firstCard.getTag().equals(figures.get(index))) {
                possibilities.add(index);
            }
            if (possibilities.size() == 2) break;
        }

        String firstCardLabel = firstCard.getResources().getResourceEntryName(firstCard.getId());
        int indexBackground = Integer.parseInt(firstCardLabel.split("d")[1]);

        if (possibilities.get(0) == indexBackground) {
            possibilities.remove(0);
        } else {
            possibilities.remove(1);
        }

        View secondCard = findViewById(getResources().getIdentifier(
                "card" + possibilities.get(0),
                "id",
                getPackageName()
        ));

        firstCard.setBackgroundResource(R.drawable.dica_animation);
        secondCard.setBackgroundResource(R.drawable.dica_animation);
        ((AnimationDrawable) firstCard.getBackground()).start();
        ((AnimationDrawable) secondCard.getBackground()).start();

        new Handler().postDelayed(() ->
                secondCard.setBackgroundResource(R.drawable.cartas), (long) (
                (AnimationDrawable) ResourcesCompat.getDrawable(getResources(),
                        R.drawable.dica_animation,
                        null)).getNumberOfFrames() * R.integer.animation_duration);
    }

    private void setButtonsClickable(boolean clickable) {
        for (int index = 0; index < 12; index++) {
            findViewById(getResources().getIdentifier(
                    "card" + index,
                    "id",
                    getPackageName()
            )).setClickable(clickable);
        }
    }
}

File: app/src/main/kotlin/com/adaca/memoria/GameActivity.kt
package com.adaca.memoria

import android.content.Intent
import android.graphics.drawable.AnimationDrawable
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.view.View
import android.view.View.*
import android.view.WindowManager
import android.widget.ImageButton
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.work.WorkManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import model.Game
import org.joda.time.DateTime
import savexport.DatabaseController
import utils.MediaPlayerManager
import utils.TimeManager
import worker.WorkerUtils

class GameActivity : AppCompatActivity() {
    private val figures = arrayListOf(
            R.drawable.circle,
            R.drawable.circle,
            R.drawable.hexagon,
            R.drawable.hexagon,
            R.drawable.losango,
            R.drawable.losango,
            R.drawable.pentagon,
            R.drawable.pentagon,
            R.drawable.triangulo,
            R.drawable.triangulo,
            R.drawable.square,
            R.drawable.square
    )

    private val timeManager = TimeManager
    private val mediaPlayerManager = MediaPlayerManager.getInstance(applicationContext)
    private val mainView = window.decorView
    private lateinit var points: TextView
    private var firstCard: ImageButton? = null

    private var hits: Int = 0
    private var tips: Int = 0
    private var misses: Int = 0
    private var missesInARow: Int = 0
    private var hit: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_game)

        timeManager.startTime = DateTime()

        figures.shuffle()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) window.attributes.layoutInDisplayCutoutMode = WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_SHORT_EDGES

        points = findViewById(R.id.tvNumPontos)
        points.text = ""
    }

    fun cardClickListener(imageButton: ImageButton) {
        if (firstCard == null || imageButton.id != firstCard?.id) {
            val label = imageButton.resources.getResourceEntryName(imageButton.id)
            val indexBackground = label.split("d")[1].toInt()
            setButtonsClickable(false)

            val figureBackground = figures[indexBackground]
            imageButton.setBackgroundResource(figureBackground)
            imageButton.tag = figureBackground

            if (firstCard != null) {
                hit = (firstCard?.tag as Int) == figureBackground
            } else {
                firstCard = imageButton
            }

            CoroutineScope(Dispatchers.Main).launch {
                delay(500).run {
                    imageButton.setBackgroundResource(R.drawable.cartas)
                    if (firstCard != imageButton) {
                        if (!hit) {
                            misses++
                            missesInARow++
                        } else {
                            mediaPlayerManager.hitFX.start()
                            points.text = (++hits).toString()
                            missesInARow = 0
                            firstCard?.visibility = View.INVISIBLE
                            imageButton.visibility = View.INVISIBLE
                            hit = false

                            if (hits == 6) {
                                startActivity(Intent(this@GameActivity, CongratulationsActivity::class.java))
                                finish()
                            }

                            if (missesInARow % 3 == 0 && missesInARow > 0) {
                                showTip(firstCard as ImageButton)
                                tips++
                            }
                            firstCard = null
                        }
                        setButtonsClickable(true)
                    }
                }
            }
        }
    }

    override fun onWindowFocusChanged(paramBoolean: Boolean) {
        super.onWindowFocusChanged(paramBoolean)

        if (paramBoolean) mainView.systemUiVisibility = (SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                or SYSTEM_UI_FLAG_LAYOUT_STABLE
                or SYSTEM_UI_FLAG_FULLSCREEN
                or SYSTEM_UI_FLAG_HIDE_NAVIGATION
                or SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                or SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION)
    }

    override fun onStop() {
        super.onStop()
        saveData()
    }

    private fun saveData() {
        timeManager.endTime = DateTime()

        val game = Game().apply {
            deviceUID = Settings.Secure.getString(contentResolver, "bluetooth_name")
            inicio = timeManager.date
            duracao = timeManager.gameTime()
            acertos = hits
            dicas = tips
            erros = misses

        }

        DatabaseController(baseContext).insertData(game).let {
            Toast.makeText(applicationContext, it, Toast.LENGTH_LONG)
        }

        WorkManager.getInstance(this).enqueue(WorkerUtils.addDataToRequest(game.toString()))
    }

    private fun showTip(firstCard: ImageButton) {
        val possibilities = arrayListOf<Int>()

        for (index in 0 until 12) {
            if (firstCard.tag.equals(figures[index])) {
                possibilities.add(index)
            }
            if (possibilities.size == 2) break
        }

        val firstCardLabel = firstCard.resources.getResourceEntryName(firstCard.id)
        val indexBackground = firstCardLabel.split("d")[1].toInt()

        if (possibilities[0] == indexBackground) {
            possibilities.remove(0)
        } else {
            possibilities.remove(1)
        }

        val secondCard: ImageButton = findViewById(resources.getIdentifier(
                "card${possibilities[0]}", "id", packageName
        ))

        firstCard.setBackgroundResource(R.drawable.dica_animation)
        secondCard.setBackgroundResource(R.drawable.dica_animation)
        (firstCard.background as AnimationDrawable).start()
        (secondCard.background as AnimationDrawable).start()
    }

    private fun setButtonsClickable(clickable: Boolean) {
        for (index in 0 until 12) {
            findViewById<ImageButton>(resources.getIdentifier(
                    "card$index",
                    "id",
                    packageName
            )).isClickable = clickable
        }
    }
}